src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStep.java

```diff
@@ -363,5 +417,8 @@ public boolean takesImplicitBlockArgument() {
         public String getWorkingDir() {
             return ContainerTemplate.DEFAULT_WORKING_DIR;
         }
+
+        public static final PodRetention defaultPodRetention = new Always();
+        public static final WorkspaceVolume defaultWorkspaceVolume = new DynamicPVCWorkspaceVolume(null, null, "ReadWriteOnce");
```

Comment:
Since you did not patch `PodTemplateStep/config.jelly`, these constants are useless.

Reply:
Also the defaults look wrong. `PodRetention.getPodTemplateDefault()` is used for `podRetention`, and it might be significant that the default is left as null. Same for `workspaceVolume` and `WorkspaceVolume getDefault()`. I suspect that the `config.jelly` defaults were broken here, but this is just making it worse but extending the problem to the runtime.

Reply:
Confirmed, the default selections in `/pipeline-syntax/` are incorrect: they do not match the actual default behavior. The generator suggests

```groovy
…, podRetention: always(), …, workspaceVolume: dynamicPVC(…), …
```

whereas the actual defaults (according to code inspection) are

```groovy
…, podRetention: default(), …, workspaceVolume: emptyDirWorkspaceVolume(false), …
```

(Note that you cannot actually use the symbol `default` as in https://github.com/jenkinsci/kubernetes-plugin/blob/a54923a10b97bad97f330923ada809e675b40c78/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pod/retention/Default.java#L65 as this is a keyword. Something else should have been chosen.)

So how does this happen?

* The dropdown under **Pod Retention** defaults to **Always** just because that is first in the list and no `default` is specified. That is simply wrong.
* The **Workspace Volume** configuration from #114 is not done in regular fashion. There should be no **Use custom workspace volume** checkbox.

It looks like #363 & #622 cleaned these things up in global pod template configuration but not in step configuration. Understandably: no one bothered to check the snippet generator, because it was so broken anyway.

Reply:
AFAICT https://github.com/jenkinsci/kubernetes-plugin/blob/a54923a10b97bad97f330923ada809e675b40c78/src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java#L899-L900 is doing the right thing for global config. https://github.com/jenkinsci/kubernetes-plugin/blob/a54923a10b97bad97f330923ada809e675b40c78/src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java#L905-L906 works in that `Default` has no fields, but is poor style; better would be

```java
public PodRetention getDefaultPodRetention() {
    return PodRetention.getPodTemplateDefault();
}
```

since `dropdownDescriptorSelector.jelly` documentation says that instances are supported for `default`:

> The default can be a specific instance or a descriptor e.g. ${descriptor.defaultSettingsProvider} or ${descriptor.defaultSettingsProvider.descriptor}. In the later case, the from input fields will be empty.

Reply:
So, as to a fix:

* I think you should revert any changes to the behavior of getters (nullability annotations are fine). The configuration here is complicated due to the layering of per-step settings over (or under?) global settings, so nulls are significant in some cases. See null checks like: https://github.com/jenkinsci/kubernetes-plugin/blob/a54923a10b97bad97f330923ada809e675b40c78/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java#L109
* The setter changes may be OK, though I am not sure. @Vlatombe would know if explicitly setting something to what ought to be its default value is actually required in order to override global settings.
* The new `defaultXXX` fields need to use the actual defaults mentioned above.
* The `config.jelly` needs to read those defaults, like `PodTemplate/config.jelly` already does.